apps/user_ldap/lib/ldapprovider.php

```
@@ -0,0 +1,76 @@
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License, version 3,
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>
+ *
+ */
+
+namespace OCA\user_ldap\lib;
+
+use OCP\LDAP\ILDAPProvider;
+use OCP\ILogger;
+use OCP\IUserManager;
+
+/**
+ * LDAP provider for pulic access to the LDAP backend.
+ */
+class LDAPProvider implements ILDAPProvider {
+
+	private $backend;
+	private $logger;
+	
+	/**
+	 * Create new LDAPProvider
+	 * @param \OCP\IUserManager $userManager
+	 * @param \OCP\ILogger $logger
+	 */
+	public function __construct(IUserManager $userManager, ILogger $logger) {
+		$this->logger = $logger;
+		foreach ($userManager->getBackends() as $backend){
+			$name = get_class($backend);
+			$this->logger->debug('instance '.$name.' backend.', ['app' => 'user_ldap']);
+			if ($backend instanceof IUserBackend && $backend->getBackendName() == USER_LDAP::BACKEND_NAME) {
+				$this->backend = $backend;
+				break;
+			}
+        }
+	}
+	
+	/**
+	 * Translate an ownCloud login name to LDAP DN
+	 * @param string $uid ownCloud user id
+	 * @throws Exception
+	 */
+	public function getUserDN($uid) {
+		if($userName = $this->backend->loginName2UserName($uid)){
+			return $this->backend->getLDAPAccess($userName)->username2dn($userName);
+		}
+		return false;
+	}
+	
+	/**
+	 * Return access for LDAP interaction for the specified user.
+	 * @param string $uid ownCloud user id
+	 * @throws Exception
+	 */
+	public function getLDAPAccess($uid) {
+		if($userName = $this->backend->loginName2UserName($uid)){
+			return $this->backend->getLDAPAccess($userName);
+		}
+		return false;
+	}
+}
```

Nope. loginnames and usernames differ. What you have often with LDAP is that a login name looks like "a.user" or "a.user@server.tld" while the username is based on the UUID which may look like "AB1234C5-678D-9E01-2E34-F567GHI89012". Basically, you'd just remove ``if($userName = $this->backend->loginName2UserName($uid)){``. 

You should always operate with usernames, not login names as they don't necessarily need to be unique. 

If your callee uses the loginname instead of the username: if the user is logged in, you'll get the username from ``\OC::$server->getUserSession()->getUser()``. This returns ``null`` or  an ``IUser`` object. In the first cast, it's an error and maybe you just want to throw an exception. In the normal case you'll get the username from  getUID() of the IUser instance.